hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateManagerV2.java

```diff
@@ -0,0 +1,88 @@
+ * <p>
+ * <p>http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * <p>Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.hadoop.hdds.scm.pipeline;
+
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.protocol.proto.HddsProtos;
+import org.apache.hadoop.hdds.scm.container.ContainerID;
+import org.apache.hadoop.hdds.scm.metadata.Replicate;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.hdds.utils.db.TableIterator;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.NavigableSet;
+
+/**
+ * Manages the state of pipelines in SCM.
+ */
+public interface PipelineStateManagerV2 {
+
+  @Replicate
+  void addPipeline(HddsProtos.Pipeline pipelineProto) throws IOException;
+
+  void addContainerToPipeline(PipelineID pipelineID,
+                              ContainerID containerID) throws IOException;
+
+  Pipeline getPipeline(PipelineID pipelineID) throws PipelineNotFoundException;
+
+  List<Pipeline> getPipelines();
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              HddsProtos.ReplicationFactor factor);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              HddsProtos.ReplicationFactor factor,
+                              Pipeline.PipelineState state);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              HddsProtos.ReplicationFactor factor,
+                              Pipeline.PipelineState state,
+                              Collection<DatanodeDetails> excludeDns,
+                              Collection<PipelineID> excludePipelines);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              Pipeline.PipelineState... states);
+
+  NavigableSet<ContainerID> getContainers(PipelineID pipelineID)
+      throws IOException;
+
+  int getNumberOfContainers(PipelineID pipelineID) throws IOException;
+
+  @Replicate
+  Pipeline removePipeline(HddsProtos.PipelineID pipelineIDProto)
+      throws IOException;
+
+  void removeContainerFromPipeline(PipelineID pipelineID,
+                                   ContainerID containerID) throws IOException;
+
+  @Replicate
+  void updatePipelineState(HddsProtos.PipelineID pipelineIDProto,
+                           HddsProtos.PipelineState newState)
+      throws IOException;
+
+  boolean isPipelineStoreEmpty() throws IOException;
+
+  TableIterator<PipelineID, ? extends Table.KeyValue<PipelineID, Pipeline>>
+      getPipelineStoreIterator() throws IOException;
```

Comment:
We should not expose `TableIterator` from `PipelineStateManager`

Reply:
Updated.